Organized Backbone Views. Edit Group works. Autopopulate works. Group lists in sync

BSteephenson opened the pull request:
This is a realllly big PR.
** This does NOT break anything (anything that wasn't broken before)

# Aims
- Organize front end so it's somewhat more manageable
- Have things sync across the page

# New Organization
- Renamed EventsView to EventsPageView, which better explains what that thing is doing
- EventsPageView has a UserView (represents the top bar thing)
- UserView has the buttons and the respective modals (contained in GroupViews.coffee and EventCRUDViews.coffee)
- GroupViews.coffee currently has AddGroupView and EditGroupView
- EventCRUDViews.coffee currently has AddEventView, EditEventView, and DeleteEventView

To picture it as an hierarchy 
- EventsPageView
    - UserView
        - AddGroupView
        - EditGroupView
        - AddEventView
        - EditEventView
        - DeleteGroupView
        - ( more to come as we do ChangeAdmins and DeleteGroup)

# Keeping Things In Sync
- EventsPageView has a UserModel
- EventsPageView passes this UserModel down to UserView
- UserView passes this UserModel down to the modals
- Since JS is pass by pointer for objects, everyone has a pointer to a singleton UserModel
- Each view can listen to events triggered by the UserModel and do the necessary DOM changes (example: @model.on 'groups.changed', @rerenderListOrSomething
- Each view makes form submissions through the UserModel

AddViewModel submits a form => UserModel makes the API call => UserModel refreshes => UserModel triggers 'groups.changed' => Everything on the page that has a list of groups makes the necessary updates

# What Works?
- Logging in and signing up
- Adding a group
- Adding an event
- Editing a group (plus auto-populate and the list is in sync) *this didn't work before*
- The Events listing stays in sync with edits to a group *this didn't work before*

# What Doesn't Work?
- Admin and Deleting a group (laziness, do it later,  I think Lenail is doing the admin part)
- Changing and Deleting an event (right now there's no straightforward way to get a user's group's events.. Need some API developments)

This isn't the perfect way to approach this, but I think this is a good base for making incremental fixes and changes in the future.

zfrenchee commented on tuftslife/assets/js/EventCRUDViews.coffee:
```diff
@@ -0,0 +1,157 @@
+			$('.tl-addEventModal').modal('hide')
+			$('.tl-addEventModal').find('input,textarea').val('')
+			# notice says "event successfuly created"
+		Event.destroy() #this might be a bad idea in the future
```
I left myself a flag here forevers ago. Do you mind resolving this before we merge? If creating a new EventModel, serializing it, and destroying it is actually the appropriate action we want to take, then that's fine. I have no idea how this is done in real life...

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,89 @@
+		@renderFilter()
+		@delegateEvents()
+		@model.on
+			"groups.changed": () => @renderGroupNames()
```
one liner

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,89 @@
+
+	initialize: ->
+		@renderFilter()
+		@delegateEvents()
```
what is @delegateEvents? it isn't in this class...

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,89 @@
+		names = _.pluck(groups, 'name')
+		@$('.tl-chgGroup-group').html('')
+		for name in names
+			console.log name
```
console logs.

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,89 @@
+		@$('.tl-chgGroup-group').html('')
+		for name in names
+			console.log name
+			@$('.tl-chgGroup-group').append("<option>#{name}</option>")
```
there's probably a more concise way to do 'renderGroupNames'. In fact I think you do this very concisely above, in the EventsView (modals) class

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,89 @@
+		#autopopulate the first one
+		if groups.length > 0
+			@selectedGroup = groups[0]
+			@populateFields(@selectedGroup)
```
this little block confuses me: 
- If the number of groups is 0, will the even be able to open the 'edit groups' modal? 
- If the number of groups is 0, is @selectedGroup null?

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,89 @@
+	populateFields : (group) ->
+		@$('.tl-chgGroup-name').val(group.name)
+		@$('.tl-chgGroup-desc').val(group.description)
+		# $('.tl-chgGroupFilters').children intersect group.tags addclass selected
```
This is going to be slightly difficult to merge with the filters edits I made, I think. I changed this commented line to a functional one in my filters PR, which is now in master. Checkout the changed I made #22 to EventsView.

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,89 @@
+		@selectedGroup = @model.getGroupWithName(groupName)
+		console.log groupName
+		console.log @selectedGroup.id
+		console.log @model.get('groups')
```
I imagine we won't need these logs?

zfrenchee commented on tuftslife/assets/js/UserModel.coffee:
```diff
@@ -1,19 +1,33 @@
-		@fetchUser()
-
 
+	# Basically a refresh function. Called (manually) after every API call
```
This seems a little bit dangerous. Can you justify this? I'm not sure how many calls we can handle.

zfrenchee commented on tuftslife/assets/js/UserModel.coffee:
```diff
@@ -1,19 +1,33 @@
 			cb?()
 
+	# An alias cuz it's shorter and makes sense in the code
+	refresh: () => @fetchUser()
```
Yeah, I can imagine we could also need to refresh other things in the future as well. good thinking.

zfrenchee commented on tuftslife/assets/js/UserModel.coffee:
```diff
@@ -36,11 +50,14 @@ class UserModel extends Backbone.Model
 
+	# Gets full name from UTLN
+	getName: () -> 
+		(@get('username').split('.').map (s) -> s.charAt(0).toUpperCase() + s.slice(1)).join(' ')
```
Why was this added up here and removed from down there? Also, no need for parens here. Functions are defined entirely by arrows, and if they have no arguments, they don't need parens

zfrenchee commented on tuftslife/assets/js/UserModel.coffee:
```diff
@@ -48,15 +65,36 @@ class UserModel extends Backbone.Model
+			return cb?(res)
 
+	editGroupAdmin: (groupname, action, username, cb) ->
+		return cb("missing field") if !(groupname? && action? && username?)
```
in coffee, I'd rather you use 'and' instead of '&&'. Also, are you sure you're DeMorgan'd here properly?

zfrenchee commented on tuftslife/assets/js/UserView.coffee:
```diff
@@ -0,0 +1,131 @@
+		if @model.get('logged.in')
+			@setupAdminTools()
+		else
+			@renderLogin()
```
I personally prefer

+		if @model.get('logged.in') then @setupAdminTools() else @renderLogin()

zfrenchee commented on tuftslife/assets/js/UserView.coffee:
```diff
@@ -0,0 +1,131 @@
+		trigger: 'click'
+		html: true
+		template: '<div class="popover" role="tooltip"><div class="arrow"></div><div class="popover-content"></div><h3 class="popover-title"></h3></div>'
+		placement: 'bottom'
```
I guess I don't like this object being a property of the view. I wish it were more hidden in a way, like an object in one of these functions here.

zfrenchee commented on tuftslife/assets/js/UserView.coffee:
```diff
@@ -0,0 +1,131 @@
+
+		popover = @$el.popover(@popOptions)
+
+		#@$el.popover('show')
```
?

zfrenchee commented on tuftslife/assets/js/UserView.coffee:
```diff
@@ -0,0 +1,131 @@
+		#@$el.popover('show')
+
+		popover.on 'shown.bs.popover', =>
+			cb?()
```
one-line

zfrenchee commented on tuftslife/assets/js/UserView.coffee:
```diff
@@ -0,0 +1,131 @@
+				if res.message is 'Account unactivated'
+					message = 'Your account seems to be unactivated. Check your email for activation instructions.'
+				if res.message is 'Unable to login'
+					message = 'Req.login failed. Check console logs'
```
I've always sort of been ashamed of this piece of code. We can fix it later, but if you have an urge to fix it now, go for it.

zfrenchee commented on tuftslife/assets/js/UserView.coffee:
```diff
@@ -0,0 +1,131 @@
+				@switchPopoverContent "Login Failed", message
+
+				hidePopover = => $(@popover).popover('hide'); @renderLogin()
+				window.setTimeout(hidePopover, 4000)	
```
This is the real shameful part though. It's a bug. No need to fix it here, but I'll fix it immediately after we merge.

zfrenchee commented on tuftslife/assets/js/UserView.coffee:
```diff
@@ -0,0 +1,131 @@
+			else
+				@$el.popover('destroy')
+				@setupAdminTools()
+				@model.on 'groups.changed', => @setupAdminTools()
```
you're binding an event here. that makes me nervous. I though you had ported all of this functionality somewhere else?

zfrenchee commented on tuftslife/tasks/pipeline.js:
```diff
@@ -57,7 +57,9 @@ var jsFilesToInject = [
-
+  'js/UserModel.js',
+  'js/GroupViews.js',
+  'js/UserView.js',
```
EventCrud? We should also rename that file. EventViews I think will do. 
Also EventPageView

zfrenchee commented on tuftslife/assets/js/UserView.coffee:
```diff
@@ -0,0 +1,131 @@
+		@model.signUp(utln, passwords[0], (res) =>
+			message = "Welcome. Check your email to complete your registration."
+
+			@switchPopoverContent('Welcome to Tuftslife!', message)
```
There are a few extra lines in here you can get rid of for clarity.

zfrenchee commented on tuftslife/assets/js/UserView.coffee:
```diff
@@ -0,0 +1,131 @@
+		if @model.get('groups')? and @model.get('groups').length > 0 
+			$('.tl-event-dropdown').removeClass('hidden')
+			@AddEventView.render()
+			@EditEventView.render()
```
I feel like we shouldn't be calling render from outside of these classes. They should probably be listening to some state, and render themselves. This is precarious.

zfrenchee commented on tuftslife/assets/js/EventsPageView.coffee:
```diff
@@ -0,0 +1,98 @@
+	initialize: ->
+		@User = new UserModel()		
+		@User.on
+			"groups.changed": @constructEvents
```
This isn't hugely clear: why would we update the events when the groups change?

zfrenchee commented on tuftslife/assets/js/EventsPageView.coffee:
```diff
@@ -0,0 +1,98 @@
+			el: $('.tl-login-button')
+			model: @User
+
+		@User.fetchUser()
```
This should be in the initialization of user, I feel like.

zfrenchee commented on tuftslife/assets/js/EventsPageView.coffee:
```diff
@@ -0,0 +1,98 @@
+
+events = new EventsPageView
+
+
```
I feel like there must be a better way to do this than all that jquery =((((((

zfrenchee commented:
I have a comment that might upset you, @BSteephenson: Do you think you could pull UserView

BSteephenson commented:
I
- made it so you only render EventsPage after (via callback) User Model fetches the user. Simplifies things a lot imo
- put back the whole Login popover code back to the way it was.

I did not
- attempt to make the tags thing work in Edit Group. See the other closed PR

BSteephenson commented:
And things render themselves. It all works

zfrenchee commented on tuftslife/assets/js/EventCRUDViews.coffee:
```diff
@@ -0,0 +1,160 @@
+			el: $('.tl-addEvent-calendar')
+			cb: (date) -> null
+
+		$('.tl-addEvent-submit-btn').click => @submitForm()
```
you could put this under 'events' (where you currently have a #TODO). That would conform better with the style you have been employing.

zfrenchee commented on tuftslife/assets/js/EventCRUDViews.coffee:
```diff
@@ -0,0 +1,160 @@
+
```
Can we rename this file from CRUD?

zfrenchee commented on tuftslife/assets/js/EventCRUDViews.coffee:
```diff
@@ -0,0 +1,160 @@
+			el: $('.tl-chgEvent-calendar')
+			cb: (date) -> null
+
+		$('.tl-chgEvent-submit-btn').click => @submitForm()
```
we can also put this under 'events' I think.

zfrenchee commented on tuftslife/assets/js/EventCRUDViews.coffee:
```diff
@@ -0,0 +1,160 @@
+		if group?
+			for event in @group.events
+				$('.tl-chgEvent-event')
+				.append("<option>#{event.name}</option>")
```
put these on the same line?

zfrenchee commented on tuftslife/assets/js/EventCRUDViews.coffee:
```diff
@@ -0,0 +1,160 @@
+		else
+			# Tell user to pick a group first
+			$('.tl-chgEvent-event')
+			.append("<option disabled>Select a Group</option>")
```
you can also one line these

zfrenchee commented on tuftslife/assets/js/EventCRUDViews.coffee:
```diff
@@ -0,0 +1,160 @@
+		endTime = moment(event.get('endTime'))
+
+		$('.tl-chgEvent-from').val(startTime.format('h:mm A'))
+		$('.tl-chgEvent-from').val(endTime.format('h:mm A'))
```
'.tl-chgEvent-to'

BSteephenson replied:
That's an oops.
Should I somehow mark in the code that this part (editing events) is still not done / not tested?

zfrenchee commented on tuftslife/assets/js/EventCRUDViews.coffee:
```diff
@@ -0,0 +1,160 @@
+			$('.tl-chgEventModal').find('input,textarea').val('')
+			# notice says "event successfuly created"
+		Event.destroy() #this might be a bad idea in the future
+		@$el.modal('hide')
```
Does this work with the backend? if you post this to editEvent, does it edit the right event?

BSteephenson replied:
That was copy paste. This whole part is untested as we don't have all the backend functionality for it (we have an edit method, we just don't know what events a user's groups' have)

zfrenchee commented on tuftslife/assets/js/EventCRUDViews.coffee:
```diff
@@ -0,0 +1,160 @@
+	submitForm: ->
+
+	renderGroupList: ->
+
```
Put 'null's at the end of each arrow. i.e. initialize: -> null

zfrenchee commented on tuftslife/assets/js/EventsPageView.coffee:
```diff
@@ -0,0 +1,98 @@
+
+		@Events = new EventsCollection
+
+		@render()
```
Are you calling render twice here? UserModel is going to fetch user, then render. Also you're calling render again. Do you need to?

BSteephenson replied:
oops

zfrenchee commented on tuftslife/assets/js/EventsPageView.coffee:
```diff
@@ -0,0 +1,98 @@
+		# Refresh the events list and content whenever groups change	
+		@User.on "groups.changed", @constructEvents
+
+		@Events = new EventsCollection
```
I would personally switch the ordering of the above two lines, I think. Semantically clearer, right?
- here's a user
- here's a bunch of events
- if the user changes the groups, update the events. 
seems like the best ordering of those statements.

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,86 @@
+	initialize: ->
+		@renderFilter()
+
+	renderFilter: ->
```
let's call this method 'render' even though right now it only renders filters.

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,86 @@
+		@renderFilter()
+		if @model.get('groups')? then @constructGroupsList()	
+
+	renderFilter: ->
```
just put this in render imo. No need for a separate function, right?

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,86 @@
+		@$('.tl-chgGroup-name').val(group.name)
+		@$('.tl-chgGroup-desc').val(group.description)
+		# TODO???
+		#@GroupTags.selectTags(group.tags)
```
as soon as we merge your PR, i'll open one which fixes this.

zfrenchee commented on tuftslife/assets/js/GroupViews.coffee:
```diff
@@ -0,0 +1,86 @@
+			description: $('.tl-chgGroup-desc').val()
+			tags: @GroupTags.active()
+		@$el.modal('hide')
+		
```
So it looks like this class can only handle the 'edit' part of the modal right now. Just wanted to clarify that that was your intention. Delete *does* have a frontend right now, but it isn't necessarily complete.

BSteephenson replied:
TODO added to the trello

zfrenchee commented on tuftslife/assets/js/UserModel.coffee:
```diff
@@ -1,19 +1,36 @@
+	initialize: (params) ->
 
+		@set('logged.in', false)
+		@fetchUser(params.afterFetch) 
```
Should this be a callback or a backbone event that someone can listen to. I think you're right that a callback is best here, but I can't say for sure. What are the pros/cons?

BSteephenson replied:
I think a callback for the first initializing is best. Reduces complexity to it.
Think of the timeline as this
fetch User => render stuff => wait for user to do something
Limiting the events pattern to that last part (UI interaction) makes life easier imo

zfrenchee commented on tuftslife/assets/js/UserModel.coffee:
```diff
@@ -1,19 +1,36 @@
 			if res.user?
 				@set('username', res.user.username)
 				@set('groups', res.user.groups)
+				if @get('logged.in') is false then @trigger 'logged.in groups.changed my.events.changed'
```
I haven't seen 'my.events.changed' used yet. Is that an event that's going to be used later or did you forget to take it out?

BSteephenson replied:
TODO for the future. Ideally, UserModel would be able to figure out what changed and emit the right events. That way you won't need to do something like Update the groups list whenever you Make an event. Right now, it's naively just updating everything.

zfrenchee commented on tuftslife/assets/js/UserModel.coffee:
```diff
@@ -48,15 +67,36 @@ class UserModel extends Backbone.Model
-
-	getName: -> (@get('username').split('.').map (s) -> s.charAt(0).toUpperCase() + s.slice(1)).join(' ')
+	editGroup: (group, cb) ->
+		return cb("No group id") if !group.id?
```
I'd prefer if you used 'not' instead of '!' and 'if (...) then return (...)' rather than 'return (...) if (...)' here (and also in the next function). What do you think?

zfrenchee commented:
Sweet. It's nearly ready to go. I'm really excited to merge this PR!

BSteephenson commented:
I
- used Backbone.View's events rather than jQuery
- made things one liners where you asked me to
- fixed the DOM class .tl-chg-Event-to thingy
- put comments indicating that EditEvent is not working and is not tested
- got rid of the second render() call in EventsPageView
- reorder method calls in EventsPageView's initialize
- gave everything a render, put renderFilter's code inside there in places
- used not instead of ! and if ... then ... instead of .... if .....
- renamed EventCrudViews to EventViews.coffee

zfrenchee commented:
YURSS!
